File: apis/sources/firms.mjs
// NASA FIRMS — Fire Information for Resource Management System
// Detects active fires/thermal anomalies within ~3 hours of satellite pass.
// Scoped to Nepal's seven provinces. Nepal's forest fire season runs roughly
// March to May, when pre-monsoon dryness combines with agricultural burning
// and fires drive both the Kathmandu valley's spring air quality collapse and
// significant losses in community forests.

import '../utils/env.mjs';
import { PROVINCES } from '../utils/nepal.mjs';

const FIRMS_BASE = 'https://firms.modaps.eosdis.nasa.gov/api/area/csv';

// Parse FIRMS CSV response into structured data
function parseCSV(rawText) {
  if (!rawText || typeof rawText !== 'string') return [];
  const lines = rawText.trim().split('\n');
  if (lines.length < 2) return [];
  const headers = lines[0].split(',');
  return lines.slice(1).map(line => {
    const vals = line.split(',');
    const obj = {};
    headers.forEach((h, i) => { obj[h.trim()] = vals[i]?.trim(); });
    return obj;
  });
}

// Fetch fires in a bounding box
async function fetchFires(opts = {}) {
  const {
    west = -180, south = -90, east = 180, north = 90,
    days = 1,
    source = 'VIIRS_SNPP_NRT',
  } = opts;

  const key = process.env.FIRMS_MAP_KEY;
  if (!key) return { error: 'No FIRMS_MAP_KEY' };

  const url = `${FIRMS_BASE}/${key}/${source}/${west},${south},${east},${north}/${days}`;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 25000);
  try {
    const res = await fetch(url, {
      signal: controller.signal,
      headers: { 'User-Agent': 'Atlas/1.0' },
    });
    clearTimeout(timer);
    if (!res.ok) return { error: `HTTP ${res.status}` };
    const text = await res.text();
    return parseCSV(text);
  } catch (e) {
    clearTimeout(timer);
    return { error: e.message };
  }
}

// Nepal's seven provinces — FIRMS wants west/south/east/north, the province
// table stores lamin/lomin/lamax/lomax, so translate once here.
const HOTSPOTS = Object.fromEntries(
  Object.entries(PROVINCES).map(([key, p]) => [
    key,
    { west: p.lomin, south: p.lamin, east: p.lomax, north: p.lamax, label: p.label },
  ])
);

// Analyze fire detections for potential military/strike activity
function analyzeFires(fires, regionLabel) {
  if (!Array.isArray(fires) || fires.length === 0) {
    return { region: regionLabel, totalDetections: 0, highConfidence: 0, highIntensity: [], summary: 'No detections' };
  }

  const highConf = fires.filter(f => f.confidence === 'h' || f.confidence === 'high');
  const nomConf = fires.filter(f => f.confidence === 'n' || f.confidence === 'nominal');

  // High intensity fires (FRP > 10 MW) — large forest fires or industrial blazes
  // rather than the small agricultural burns that dominate detection counts
  const highIntensity = fires
    .filter(f => parseFloat(f.frp) > 10)
    .map(f => ({
      lat: parseFloat(f.latitude),
      lon: parseFloat(f.longitude),
      brightness: parseFloat(f.bright_ti4),
      frp: parseFloat(f.frp),
      date: f.acq_date,
      time: f.acq_time,
      confidence: f.confidence,
      daynight: f.daynight,
    }))
    .sort((a, b) => b.frp - a.frp)
    .slice(0, 15);

  // Night detections matter most: crop-residue and pasture burning happens in
  // daylight, so a night signature usually means a fire nobody put out
  const nightFires = fires.filter(f => f.daynight === 'N');

  return {
    region: regionLabel,
    totalDetections: fires.length,
    highConfidence: highConf.length,
    nominalConfidence: nomConf.length,
    nightDetections: nightFires.length,
    highIntensity,
    avgFRP: fires.reduce((sum, f) => sum + (parseFloat(f.frp) || 0), 0) / fires.length,
  };
}

// Briefing
export async function briefing() {
  const key = process.env.FIRMS_MAP_KEY;
  if (!key) {
    return {
      source: 'NASA FIRMS',
      timestamp: new Date().toISOString(),
      status: 'no_key',
      message: 'Set FIRMS_MAP_KEY for satellite fire/strike detection. Free at https://firms.modaps.eosdis.nasa.gov/api/area/',
    };
  }

  // Fetch all hotspots in parallel
  const entries = Object.entries(HOTSPOTS);
  const rawResults = await Promise.all(
    entries.map(async ([key, box]) => {
      const fires = await fetchFires({ ...box, days: 2 });
      return { key, label: box.label, fires };
    })
  );

  const hotspots = rawResults.map(r => {
    if (r.fires?.error) return { region: r.label, error: r.fires.error };
    return analyzeFires(r.fires, r.label);
  });

  // Generate signals
  const signals = [];
  const totalDetections = hotspots.reduce((s, h) => s + (h.totalDetections || 0), 0);

  for (const h of hotspots) {
    if (h.highIntensity?.length > 5) {
      signals.push(`HIGH INTENSITY FIRES in ${h.region}: ${h.highIntensity.length} detections >10MW FRP`);
    }
    if (h.nightDetections > 20) {
      signals.push(`SUSTAINED NIGHT BURNING in ${h.region}: ${h.nightDetections} night detections — fires running unchecked overnight`);
    }
  }

  // Pre-monsoon is when Nepal's fire season peaks and smoke drives the
  // Kathmandu valley's worst air quality of the year.
  const month = new Date().getMonth() + 1;
  const fireSeason = month >= 3 && month <= 5;
  if (fireSeason && totalDetections > 200) {
    signals.push(`Fire season active — ${totalDetections} detections nationwide, expect degraded air quality across the valley`);
  }

  return {
    source: 'NASA FIRMS',
    timestamp: new Date().toISOString(),
    status: 'active',
    fireSeason,
    totalDetections,
    hotspots,
    signals,
  };
}

if (process.argv[1]?.endsWith('firms.mjs')) {
  const data = await briefing();
  console.log(JSON.stringify(data, null, 2));
}

File: backend/app/domains/hazards/sources/firms.py
"""NASA FIRMS — active fires and thermal anomalies over Nepal.

Detects fires within roughly three hours of a satellite pass, scoped to the
seven provinces. Nepal's forest fire season runs roughly March to May, when
pre-monsoon dryness combines with agricultural burning — and that smoke drives
the Kathmandu valley's spring air quality collapse.

Needs FIRMS_MAP_KEY. Without it this reports status 'no_key' and the wildfire
panel is empty, which is the correct degraded state.

Runnable alone:  python -m app.domains.hazards.sources.firms
"""

import asyncio
import csv
import io
from datetime import datetime, timezone
from typing import Any

from app.core.config import settings
from app.core.http import is_error, now_iso, safe_fetch
from app.core.nepal import PROVINCES

FIRMS_BASE = "https://firms.modaps.eosdis.nasa.gov/api/area/csv"
DEFAULT_SOURCE = "VIIRS_SNPP_NRT"

# FIRMS wants west/south/east/north; the province table stores
# lamin/lomin/lamax/lomax. Translate once, here.
HOTSPOTS = {
    key: {
        "west": p.lomin,
        "south": p.lamin,
        "east": p.lomax,
        "north": p.lamax,
        "label": p.label,
    }
    for key, p in PROVINCES.items()
}


def parse_csv(raw: str) -> list[dict[str, str]]:
    """FIRMS answers CSV, not JSON."""
    if not raw or not isinstance(raw, str):
        return []
    rows = list(csv.DictReader(io.StringIO(raw.strip())))
    return [{(k or "").strip(): (v or "").strip() for k, v in row.items()} for row in rows]


def _as_float(value: Any) -> float | None:
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


async def fetch_fires(
    west: float, south: float, east: float, north: float, days: int = 1
) -> Any:
    key = settings.FIRMS_MAP_KEY
    if not key:
        return {"error": "No FIRMS_MAP_KEY"}

    url = f"{FIRMS_BASE}/{key}/{DEFAULT_SOURCE}/{west},{south},{east},{north}/{days}"
    result = await safe_fetch(url, timeout=25.0, retries=0, as_="text")
    if is_error(result):
        return {"error": result.error}
    return parse_csv(result)


def analyze_fires(fires: list[dict[str, str]], region_label: str) -> dict[str, Any]:
    if not fires:
        return {
            "region": region_label,
            "totalDetections": 0,
            "highConfidence": 0,
            "highIntensity": [],
            "summary": "No detections",
        }

    high_conf = [f for f in fires if f.get("confidence") in ("h", "high")]
    nominal_conf = [f for f in fires if f.get("confidence") in ("n", "nominal")]

    # FRP > 10 MW: large forest fires or industrial blazes, rather than the
    # small agricultural burns that dominate raw detection counts.
    high_intensity = sorted(
        (
            {
                "lat": _as_float(f.get("latitude")),
                "lon": _as_float(f.get("longitude")),
                "brightness": _as_float(f.get("bright_ti4")),
                "frp": _as_float(f.get("frp")),
                "date": f.get("acq_date"),
                "time": f.get("acq_time"),
                "confidence": f.get("confidence"),
                "daynight": f.get("daynight"),
            }
            for f in fires
            if (_as_float(f.get("frp")) or 0) > 10
        ),
        key=lambda f: f["frp"] or 0,
        reverse=True,
    )[:15]

    # Night detections matter most: crop-residue and pasture burning happens in
    # daylight, so a night signature usually means a fire nobody put out.
    night_fires = [f for f in fires if f.get("daynight") == "N"]

    return {
        "region": region_label,
        "totalDetections": len(fires),
        "highConfidence": len(high_conf),
        "nominalConfidence": len(nominal_conf),
        "nightDetections": len(night_fires),
        "highIntensity": high_intensity,
        "avgFRP": sum(_as_float(f.get("frp")) or 0 for f in fires) / len(fires),
    }


def is_fire_season(when: datetime | None = None) -> bool:
    """Pre-monsoon, March to May, is when Nepal's fire season peaks."""
    month = (when or datetime.now(timezone.utc)).month
    return 3 <= month <= 5


async def _for_region(key: str, box: dict[str, Any]) -> dict[str, Any]:
    fires = await fetch_fires(
        west=box["west"], south=box["south"], east=box["east"], north=box["north"], days=2
    )
    if isinstance(fires, dict) and fires.get("error"):
        return {"region": box["label"], "error": fires["error"]}
    return analyze_fires(fires, box["label"])


async def briefing() -> dict[str, Any]:
    if not settings.FIRMS_MAP_KEY:
        return {
            "source": "NASA FIRMS",
            "timestamp": now_iso(),
            "status": "no_key",
            "message": (
                "Set FIRMS_MAP_KEY for satellite fire detection. Free at "
                "https://firms.modaps.eosdis.nasa.gov/api/area/"
            ),
        }

    hotspots = list(
        await asyncio.gather(*(_for_region(k, box) for k, box in HOTSPOTS.items()))
    )

    signals: list[str] = []
    total_detections = sum(h.get("totalDetections") or 0 for h in hotspots)

    for h in hotspots:
        if len(h.get("highIntensity") or []) > 5:
            signals.append(
                f"HIGH INTENSITY FIRES in {h['region']}: "
                f"{len(h['highIntensity'])} detections >10MW FRP"
            )
        if (h.get("nightDetections") or 0) > 20:
            signals.append(
                f"SUSTAINED NIGHT BURNING in {h['region']}: "
                f"{h['nightDetections']} night detections — fires running unchecked overnight"
            )

    fire_season = is_fire_season()
    if fire_season and total_detections > 200:
        signals.append(
            f"Fire season active — {total_detections} detections nationwide, "
            "expect degraded air quality across the valley"
        )

    return {
        "source": "NASA FIRMS",
        "timestamp": now_iso(),
        "status": "active",
        "fireSeason": fire_season,
        "totalDetections": total_detections,
        "hotspots": hotspots,
        "signals": signals,
    }


if __name__ == "__main__":
    import json

    print(json.dumps(asyncio.run(briefing()), indent=2, ensure_ascii=False))
